hail/src/main/scala/is/hail/expr/ir/StagedRegion.scala

```diff
@@ -0,0 +1,74 @@
+package is.hail.expr.ir
+
+import is.hail.annotations.Region
+import is.hail.asm4s.{Code, Settable, Value}
+import is.hail.types.physical.PCode
+import is.hail.utils._
+
+abstract class StagedRegion {
+  def code: Value[Region]
+
+  def createChildRegion(mb: EmitMethodBuilder[_]): StagedOwnedRegion
+}
+
+trait StagedOwnedRegion extends StagedRegion {
+  def allocateRegion(size: Int): Code[Unit]
+
+  def free(): Code[Unit]
+
+  def clear(): Code[Unit]
+
+  def giveToParent(): Code[Unit]
+
+  def shareWithParent(): Code[Unit]
+
+  def copyToParent(mb: EmitMethodBuilder[_], value: PCode): PCode
+}
+
+object StagedRegion {
+  def apply(r: Value[Region], allowSubregions: Boolean = false): StagedRegion =
+    if (allowSubregions) new RealStagedRegion(r) else new DummyStagedRegion(r)
+
+  class RealStagedRegion(r: Value[Region]) extends StagedRegion {
```

Comment:
Prefer not to define this in the object, the compiled class's name gets mangled and this has been annoying in the past.

Reply:
Ah, sure.